src/System.Linq/src/System/Linq/Partition.cs

```diff
@@ -42,22 +42,18 @@ internal interface IPartition<TElement> : IIListProvider<TElement>
 
         IPartition<TElement> Take(int count);
 
-        TElement ElementAt(int index);
+        TElement TryGetElementAt(int index, out bool found);
 
-        TElement ElementAtOrDefault(int index);
+        TElement TryGetFirst(out bool found);
 
-        TElement First();
-
-        TElement FirstOrDefault();
-
-        TElement Last();
-
-        TElement LastOrDefault();
+        TElement TryGetLast(out bool found);
```

Comment:
Yes, that is for Ts that are references. JIT must conservatively check whether out param is not pointing  to a heap location, since in such case it could be a cross-generational store (if Gen0 object is referenced from Gen1 object, it is must be considered rooted for the purpose of Gen0 GC).

I had cases where swapping bool and return value in TryGet was guided by profiling pointing at checked write barrier being nontrivial expense.
like here: https://github.com/VSadov/NonBlocking/commit/7fc7e88ebac8806fd16e8f73d3c4b83a7aa6eb78